File: trading-analysis-platform/backend/app/services/macro/macro_types.py
"""Tipos y helpers del Macro Dashboard."""
from __future__ import annotations

# Trend
IMPROVING = "IMPROVING"
WORSENING = "WORSENING"
STABLE = "STABLE"
UNKNOWN = "UNKNOWN"

# Status
POSITIVE = "POSITIVE"
NEUTRAL = "NEUTRAL"
NEGATIVE = "NEGATIVE"
MISSING = "MISSING"

# Risk
GREEN = "GREEN"
YELLOW = "YELLOW"
RED = "RED"
RISK_UNKNOWN = "UNKNOWN"

# Yield curve
CURVE_NORMAL = "NORMAL"
CURVE_FLAT = "FLAT"
CURVE_INVERTED = "INVERTED"
CURVE_UNKNOWN = "UNKNOWN"


def indicator(
    key: str,
    label: str,
    value: float | None,
    *,
    display_value: str | None = None,
    previous_value: float | None = None,
    change: float | None = None,
    trend: str = UNKNOWN,
    status: str = NEUTRAL,
    source: str = "FRED",
    last_updated: str | None = None,
    explanation: str | None = None,
    extra: dict | None = None,
) -> dict:
    out = {
        "key": key,
        "label": label,
        "value": value,
        "displayValue": display_value if display_value is not None else (
            "Unavailable" if value is None else str(value)
        ),
        "previousValue": previous_value,
        "change": change,
        "trend": trend,
        "status": status,
        "source": source,
        "lastUpdated": last_updated,
        "explanation": explanation,
    }
    if extra:
        out.update(extra)
    return out


def missing_indicator(key: str, label: str, source: str = "FRED") -> dict:
    return indicator(
        key, label, None,
        display_value="Unavailable", trend=UNKNOWN, status=MISSING, source=source,
        explanation="Provider not configured or data unavailable.",
    )

File: trading-analysis-platform/frontend/src/features/macro/macroTypes.ts
// Tipos del Macro Dashboard (forma de /api/macro/overview).
// Informativo: no es señal de compra/venta. Datos parciales si faltan proveedores.

export type MacroStatus = "POSITIVE" | "NEUTRAL" | "NEGATIVE" | "MISSING";
export type MacroTrend = "IMPROVING" | "WORSENING" | "STABLE" | "UNKNOWN";
export type MacroRiskLevel = "GREEN" | "YELLOW" | "RED" | "UNKNOWN";
export type YieldCurveStatus = "NORMAL" | "FLAT" | "INVERTED" | "UNKNOWN";

export interface MacroIndicator {
  key: string;
  label: string;
  value: number | null;
  displayValue: string;
  previousValue?: number | null;
  change?: number | null;
  trend: MacroTrend;
  status: MacroStatus;
  source: string;
  lastUpdated?: string | null;
  explanation?: string | null;
  /** Clave para el texto de ayuda del tooltip (por defecto = key). */
  helpKey?: string;
  // Extra para instrumentos de mercado (FX/commodities/crypto).
  symbol?: string;
  changePercent?: number | null;
}

export interface EconomicCalendarEvent {
  eventName: string;
  date: string | null;
  time?: string | null;
  country?: string | null;
  impact: "HIGH" | "MEDIUM" | "LOW";
  consensus?: string | null;
  previous?: string | null;
  source: string;
  notes?: string | null;
}

export interface MacroRisk {
  riskLevel: MacroRiskLevel;
  score: number | null;
  drivers: string[];
  risks: string[];
}

export interface MacroExecutiveSummary {
  riskLevel: MacroRiskLevel;
  riskLabel: string;
  summary: string;
  lastUpdated: string | null;
}

export interface MacroRates {
  treasury2Y?: MacroIndicator;
  treasury5Y?: MacroIndicator;
  treasury10Y?: MacroIndicator;
  treasury30Y?: MacroIndicator;
  yieldCurve10Y2Y?: MacroIndicator;
  curveStatus: YieldCurveStatus;
  [key: string]: MacroIndicator | YieldCurveStatus | undefined;
}

export interface MacroOverviewResponse {
  executiveSummary: MacroExecutiveSummary;
  macroRisk: MacroRisk;
  usaIndicators: Record<string, MacroIndicator>;
  rates: MacroRates;
  globalMarkets: {
    fx: MacroIndicator[];
    commodities: MacroIndicator[];
    crypto: MacroIndicator[];
  };
  economicCalendar: EconomicCalendarEvent[];
  economicCalendarAvailable?: boolean;
  economicCalendarSource?: "FRED" | "CONFIGURED_PROVIDER" | "UNAVAILABLE";
  whatThisMeans: string[];
  dataAvailability: Record<string, boolean>;
  warnings: string[];
  lastUpdated: string | null;
  fromCache?: boolean;
}

export const RISK_COLOR: Record<MacroRiskLevel, string> = {
  GREEN: "#22c55e",
  YELLOW: "#f59e0b",
  RED: "#ef4444",
  UNKNOWN: "#6b7280",
};

export const RISK_LABEL_ES: Record<MacroRiskLevel, string> = {
  GREEN: "Riesgo bajo",
  YELLOW: "Riesgo moderado",
  RED: "Riesgo elevado",
  UNKNOWN: "Sin datos",
};

export const CURVE_LABEL_ES: Record<YieldCurveStatus, string> = {
  NORMAL: "Normal",
  FLAT: "Plana",
  INVERTED: "Invertida",
  UNKNOWN: "Sin datos",
};

export const STATUS_TEXT_COLOR: Record<MacroStatus, string> = {
  POSITIVE: "text-up",
  NEUTRAL: "text-muted",
  NEGATIVE: "text-down",
  MISSING: "text-muted",
};
